rpm/dnf/libdnf_wrapper.cpp

```diff
@@ -0,0 +1,242 @@
+// +build dnf
+
+#include "libdnf_wrapper.h"
+#include <exception>
+#include <string>
+#include <string.h>
+
+const char* newCString(std::string s) {
+    char* msg = (char*) malloc(s.size()+1);
+    strcpy(msg, s.c_str());
+    return msg;
+}
+
+const char* getErrorMessage(GError* gerr) {
+    if (gerr == nullptr) {
+        return newCString("unknown error");
+    }
+
+    const char* msg = strdup(gerr->message);
+    g_error_free(gerr);
+    return msg;
+}
+
+const char* getErrorMessage(std::string prefix, GError* gerr) {
+    if (gerr == nullptr) {
+        return newCString(prefix + "unknown error");
+    }
+
+    const char* msg = newCString(prefix + std::string(gerr->message));
+    g_error_free(gerr);
+    return msg;
+}
+
+const char* SetupDNFSack(DnfContext* context) {
+    try {
+        DnfState* state = dnf_context_get_state(context);
+
+        GError* gerr = nullptr;
+        if (dnf_context_setup_sack(context, state, &gerr) == 0) {
+            return getErrorMessage(gerr);
+        }
+    } catch(std::exception &e) {
+        return strdup(e.what());
+    }
+    return nullptr;
+}
+
+LookupPackageResult LookupPackage(DnfContext* context, int filter, int comparison, const char* value) {
+    LookupPackageResult result = {0};
+    HyQuery query = nullptr;
+    GPtrArray* pkglist = nullptr;
+    try {
+        DnfSack* sack = dnf_context_get_sack(context);
+        query = hy_query_create(sack);
+        hy_query_filter(query, filter, comparison, value);
+        pkglist = hy_query_run(query);
+
+        if (pkglist == nullptr || pkglist->len == 0) {
+            result.err_msg = newCString("failed to find package");
+        } else {
+            result.pkg = static_cast<DnfPackage*>(g_object_ref(g_ptr_array_index(pkglist, 0)));
+        }
+    } catch(std::exception &e) {
+        result.err_msg = strdup(e.what());
```

Comment:
`e.what()` returns a `const char*` while `newCString` accepts a `std::string`, so I just thought it was slightly nicer to avoid constructing a new string by using `strdup` instead 🤷‍♀️  but I really don't feel strongly about it so if you think `newCString` would be more appropriate I can switch it